### PR TITLE
PaginatedDataTable (part 1)

### PR DESCRIPTION
**Pull request by Hixie**
This introduces the key parts of a paginated data table, not including
the built-in pagination features.
- Provide more data for the data table demo, so there's data to page.
- Introduce a ChangeNotifier class which abstracts out
  addListener/removeListener/notifyListeners. We might be able to use
  this to simplify existing classes as well, though this patch doesn't
  do that.
- Introduce DataTableSource, a delegate for getting data for data
  tables. This will also be used by ScrollingDataTable in due course.
- Introduce PaginatedDataTable, a widget that wraps DataTable and only
  shows N rows at a time, fed by a DataTableSource.

### CODE REVIEW

**Comment by Hixie** [comment]
This patch was growing too big. I'm tired of having long-lived branches, so I figured I'd submit this and then get back to adding the back/forward UI in a separate patch.

**Comment by Hixie** [comment]
@abarth

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+abstract class ChangeNotifier {
+  List<VoidCallback> _listeners;
+  void addListener(VoidCallback listener) {
+    _listeners ??= <VoidCallback>[];
+    _listeners.add(listener);
+  }
+  void removeListener(VoidCallback listener) {
+    _listeners?.remove(listener);
+  }
+  void dispose() {
```
Comment (abarth) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: `@mustCallSuper`

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// found in the LICENSE file.
+
+import 'assertions.dart';
+import 'basic_types.dart';
+
+/// Abstract class that can be extended or mixed in that provides
+/// a change notification API using [VoidCallback] for notifications.
+abstract class ChangeNotifier {
+  List<VoidCallback> _listeners;
+  void addListener(VoidCallback listener) {
```
Comment (abarth) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: Please add dartdocs

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    _listeners ??= <VoidCallback>[];
+    _listeners.add(listener);
+  }
+  void removeListener(VoidCallback listener) {
+    _listeners?.remove(listener);
+  }
+  void dispose() {
+    _listeners = null;
+  }
+  void notifyListeners() {
```
Comment (abarth) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: `@protected` ?

File: packages/flutter/lib/src/foundation/change_notifier.dart [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    if (_listeners != null) {
+      for (VoidCallback listener in _listeners) {
```
Comment (abarth) [inline packages/flutter/lib/src/foundation/change_notifier.dart]: We should make a local copy of the iterators while iterating them.

**Comment by abarth** [comment]
LGTM

**Comment by Hixie** [comment]
fixed all comments and added a test for change notifier.

**Pull request merged**